### PR TITLE
Auth: Add tls_client_cert to OAuth section in defaults.ini. #64746

### LINKED ISSUE #64746: Allow further variable overrides in gitlab_oauth
Further secure OAuth authentication with gitlab by allowing further variable overrides of gitlab_oauth

- see also
  - https://github.com/grafana/grafana/issues/64648#issuecomment-1466362388
  - https://github.com/grafana/grafana/pull/64649#issuecomment-1467150395

# Functional example

This works with custom builds, but I would like to see it natively available.

## variables

```
      GF_AUTH_GITLAB_ALLOWED_DOMAINS: example.com
      GF_AUTH_GITLAB_TLS_CLIENT_CA: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_CERT: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_KEY: /run/secrets/gitlab-key
      GF_AUTH_GITLAB_USE_PKCE: true
```

## docker　compose example

```
#################################### GitLab Auth #########################
[auth.gitlab]
#~~~~
#new
allowed_domains = example.com
tls_skip_verify_insecure = false
tls_client_cert =
tls_client_key =
tls_client_ca =
use_pkce = false

```

```Dockerfile
# syntax=docker/dockerfile:1.5

FROM grafana/grafana:9.4.3

COPY ./conf/defaults.ini /usr/share/grafana/conf/defaults.ini
```

```yaml
services:
  grafana:
    build: .
    #image: 'grafana/grafana:9.4.3'
    hostname: grafana.example.com
    restart: always
    environment:
      GF_SERVER_ROOT_URL: https://grafana.example.com
      GF_SECURITY_ADMIN_USER: admin
      GF_SECURITY_ADMIN_PASSWORD: Grafana
      GF_AUTH_GITLAB_ENABLED: true
      GF_AUTH_GITLAB_ALLOW_SIGNUP: true
      GF_AUTH_GITLAB_CLIENT_ID: 194d59efb736621770baa45dd8a9a40689a24854fc4e99b74a869a863d8690c2
      GF_AUTH_GITLAB_CLIENT_SECRET: fce2b0c710a99cc9201d1687d9b38b7ddbb3dc4e6714a425e6bb78afcf37e599
      GF_AUTH_GITLAB_ROLE_ATTRIBUTE_PATH: is_admin && 'Admin' && 'GrafanaAdmin' || 'Viewer'
      GF_AUTH_GITLAB_SCOPES: read_user,read_api
      GF_AUTH_GITLAB_AUTH_URL:  https://gitlab.example.com/oauth/authorize
      GF_AUTH_GITLAB_TOKEN_URL: https://gitlab.example.com/oauth/token
      GF_AUTH_GITLAB_API_URL:   https://gitlab.example.com/api/v4
      GF_AUTH_GITLAB_ALLOW_ASSIGN_GRAFANA_ADMIN: true
      GF_AUTH_GITLAB_TLS_SKIP_VERIFY_INSECURE: false
      # some-secure-settings
      GF_AUTH_GITLAB_ALLOWED_DOMAINS: example.com
      GF_AUTH_GITLAB_TLS_CLIENT_CA: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_CERT: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_KEY: /run/secrets/gitlab-key
      GF_AUTH_GITLAB_USE_PKCE: true
    extra_hosts:
      - "gitlab.example.com:192.168.***.***"
    volumes:
      - grafana-config:/bin
      - grafana:/var/lib/grafana
    secrets:
      - gitlab-cert
      - gitlab-key
```

### trust self-signed cert

```
      GF_AUTH_GITLAB_TLS_SKIP_VERIFY_INSECURE: false
      GF_AUTH_GITLAB_TLS_CLIENT_CA: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_CERT: /run/secrets/gitlab-cert
      GF_AUTH_GITLAB_TLS_CLIENT_KEY: /run/secrets/gitlab-key
```

### enable PKCE

`GF_AUTH_GITLAB_USE_PKCE: true`

```
Started GET "/oauth/authorize?client_id=194d59efb736621770baa45dd8a9a40689a24854fc4e99b74a869a863d8690c2&code_challenge=eJmtACxjbVjm0lW-PSreMb5p_Nakys9ytKF-jrinNww&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fgrafana.aruki.com%2Flogin%2Fgitlab&response_type=code&scope=read_user+read_api&state=_MKwiuf4gdQL-FfiGfpkQ9-qxYv72DX69hg3ju0_NO4%3D" for 172.19.0.2 at 2023-03-14 14:31:41 +0000
Processing by Oauth::AuthorizationsController#new as HTML
  Parameters: {"client_id"=>"194d59efb736621770baa45dd8a9a40689a24854fc4e99b74a869a863d8690c2", "code_challenge"=>"eJmtACxjbVjm0lW-PSreMb5p_Nakys9ytKF-jrinNww", "code_challenge_method"=>"S256", "redirect_uri"=>"https://grafana.aruki.com/login/gitlab", "response_type"=>"code", "scope"=>"read_user read_api", "state"=>"_MKwiuf4gdQL-FfiGfpkQ9-qxYv72DX69hg3ju0_NO4="}
  Rendered layout layouts/profile.html.haml (Duration: 604.3ms | Allocations: 254654)
Completed 200 OK in 766ms (Views: 561.3ms | ActiveRecord: 121.2ms | Elasticsearch: 0.0ms | Allocations: 273206)
```

### limits access to specific domains.(GF_AUTH_GITLAB_ALLOWED_DOMAINS)

`GF_AUTH_GITLAB_ALLOWED_DOMAINS: example.com`

image

# refer

Thanks your review, I got it.
Expand a bit,I want to add trust specific self-signed certificate, add enable [PKCE](https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce) 

- GF_AUTH_GITLAB_TLS_CLIENT_CERT
- GF_AUTH_GITLAB_TLS_CLIENT_KEY
- GF_AUTH_GITLAB_TLS_CLIENT_CA
- GF_AUTH_GITLAB_USE_PKCE

_Originally posted by @arukiidou in https://github.com/grafana/grafana/issues/64649#issuecomment-1467150395_

### PR DESCRIPTION
**Pull request by arukiidou**
**What is this feature?**

- Add `tls_client_cert` to defaults.ini file for oauth providers. This enabled the ability to override these settings with env variables.
  - `tls_client_cert`
  - `tls_client_key` 
  - `tls_client_ca `

**Why do we need this feature?**

- This is safer than `tls_skip_verify_insecure`
  - follow-up 
    - https://github.com/grafana/grafana/pull/64649
    - https://github.com/grafana/grafana/pull/64679

**Who is this feature for?**

- Primarily GitLab user.
  - Previously, GitLab bundled Grafana to use OAuth out of the box
  - Grafana bundle will deprecated in GitLab 16.0.
  - It is possible that GitLab users will start using Grafana as well if TLS ease of use is improved.

**Which issue(s) does this PR fix?**:

- Fixes #64746

**Special notes for your reviewer:**

- suggest labels
  - https://github.com/grafana/grafana/labels/area%2Fauth%2Foauth
  - https://github.com/grafana/grafana/labels/no-changelog
  - https://github.com/grafana/grafana/labels/no-backport

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: docs/sources/setup-grafana/configure-security/configure-authentication/gitlab/index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/gitlab/index.md:94]:
```suggestion
`tls_skip_verify_insecure` controls whether a client verifies the server's certificate chain and host name. If set to `true`, then SSL/TLS accepts any certificate the server presents and any host name in that certificate. _You should only use this for testing_because this mode leaves SSL/TLS susceptible to man-in-the-middle attacks.
```

**Review by chri2547 — COMMENTED** [review]
A small docs nit, approving so as not to block.

**Comment by arukiidou** [comment]
There is a docs-lint error, I will fix it today

**Comment by arukiidou** [comment]
> A small docs nit, approving so as not to block.

- I would like to revert the revision of your suggest for two reasons
  - I have tried your suggestion but cannot clear the prettier-linter
  - The original description is the same as [Generic OAuth](https://github.com/grafana/grafana/blob/main/docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth/index.md)

**Comment by arukiidou** [comment]
- Resolve conflicts brought from #68095.

**Comment by chri2547** [comment]
Docs soft approval given, waiting on Engr for review and formal approval.

**Comment by chri2547** [comment]
@grafana/backend-platform , could I please have some eyes on this PR? Thank you.

**Comment by marefr** [comment]
Functionality introduced 7 years ago via https://github.com/grafana/grafana/pull/6226 :smile: @grafana/grafana-authnz-team please have a look at this

**Comment by arukiidou** [comment]
Thanks to the maintainers for moving this forward!

**Review by IevaVasiljeva — APPROVED** [review]
Looks good to me, thanks for the contribution!

I've also added a `no-backport` label to make the label checks pass. Let me know if you think that this PR should be backported, and we can discuss it.

**Pull request merged**
